rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala

```diff
@@ -39,21 +39,23 @@ import monix.eval.Coeval
   */
 trait Reduce[M[_]] {
 
-  def produce(chan: Quote, data: Seq[Par], persistent: Boolean)(implicit env: Env[Par],
-                                                                rand: Blake2b512Random): M[Unit]
-
-  def consume(binds: Seq[(BindPattern, Quote)], body: Par, persistent: Boolean)(
-      implicit env: Env[Par],
-      rand: Blake2b512Random): M[Unit]
+  protected[this] implicit def monadM: Monad[M]
 
   def eval(par: Par)(implicit env: Env[Par], rand: Blake2b512Random): M[Unit]
 
-  def inj(par: Par)(implicit rand: Blake2b512Random): M[Unit]
+  def inj(par: Par)(implicit rand: Blake2b512Random): M[Unit] =
+    for { _ <- eval(par)(Env[Par](), rand) } yield ()
 
   /**
     * Evaluate any top level expressions in @param Par .
     */
-  def evalExpr(par: Par)(implicit env: Env[Par]): M[Par]
+  def evalExpr(par: Par)(implicit env: Env[Par]): M[Par] =
+    for {
+      evaledExprs <- par.exprs.toList.traverse(expr => evalExprToPar(expr))
+      result = evaledExprs.foldLeft(par.copy(exprs = Vector())) { (acc, newPar) =>
+        acc ++ newPar
+      }
+    } yield result
```

Comment:
Implementations of those methods were moved out from the `trait` on purpose. I plann on removing `evalExpr`  and `evalExprToPar` from the `trait` alltogether leaving only `inj` and `eval`.

Reply:
Great! :) I've dropped that commit. Good to merge?